lib/fluent/plugin/parser.rb

```diff
@@ -17,13 +17,58 @@
 require 'fluent/plugin/base'
 require 'fluent/plugin/owned_by_mixin'
 
+require 'fluent/error'
 require 'fluent/mixin' # for TypeConverter
 require 'fluent/time'
 require 'fluent/plugin/string_util'
 
+require 'serverengine/blocking_flag'
+
 module Fluent
   module Plugin
     class Parser < Base
+      class TimeoutChecker
+        # This implementation now uses mutex because parser is typically used in input.
+        # If this has a performance issue under high concurreny, use concurrent-ruby's map instead.
+        def initialize(timeout)
+          @map = {}
+          @flag = ServerEngine::BlockingFlag.new
+          @mutex = Mutex.new
+          @timeout = timeout
+        end
+
+        def start
+          @thread = ::Thread.new {
+            until @flag.wait_for_set(0.5)
+              @mutex.synchronize {
+                now = Time.now
+                @map.keys.each { |th|
+                  time = @map[th]
+                  if now - time > @timeout
+                    th.raise UnrecoverableError, "parsing timed out"
```

Comment:
it's better to use `TimeoutError` or like that since `UnrecoverableError` shouldn't be used for recoverable error.

Reply:
This aims to avoid using 3rd party exception to incorrect catch. Another idea is create new exception for it.

Reply:
> Another idea is create new exception for it.

this looks good.